tools/daemon/flow.c

```diff
@@ -0,0 +1,562 @@
+			free(cur_element);
+			rc = -EFAULT;
+			goto err;
+		}
+
+		INIT_LIST_HEAD(&cur_element->list);
+		cur_element->ref = 0;
+		cur_element->value[0] = value->if_id;
+		cur_element->ctx = (void *)calloc(1, sizeof(*cur_element->ctx));
+		if (NULL == cur_element->ctx) {
+			free(cur_element);
+			rc = -ENOMEM;
+			goto err;
+		}
+		/* tables from 0x800 are reserved by kernel */
+		bitmap_create(&cur_element->ctx->ht, (0x800 - 1));
+		if (NULL == cur_element->ctx->ht) {
+			free(cur_element->ctx);
+			free(cur_element);
+			rc = -ENOMEM;
+			goto err;
+		}
+
+		/* table id = 0 is not used */
+		bitmap_set(cur_element->ctx->ht, 0);
+		list_add_tail(&cur_element->item, cur_head);
+	}
+	assert(cur_element);
+	cur_element->ref++;
+	ctx = cur_element->ctx;
+
+	log_debug("[%d] add flow (if): 0x%p value: %d ref: %d\n",
+			pid, cur_element, cur_element->value[0], cur_element->ref);
+
+	/* ip list processing */
+	cur_head = &cur_element->list;
+	list_for_each(cur_entry, cur_head) {
+		cur_element = container_of(cur_entry, struct flow_element, item);
+		if (cur_element->value[0] == (uint32_t)value->type &&
+			cur_element->value[1] == ip) {
+			ht = cur_element->ht_id & 0x0000FFFF;
+			ht_internal = (cur_element->ht_id >> 16) & 0x0000FFFF;
+			break;
+		}
+	}
+	if (cur_entry == cur_head) {
+		cur_element = (void *)calloc(1, sizeof(*cur_element));
+		if (NULL == cur_element) {
+			rc = -ENOMEM;
+			goto err;
+		}
+
+		get_htid(ctx, get_prio(value), &ht_internal, &ht);
+		sys_exec("tc filter del dev %s parent ffff: protocol ip prio %d handle %x: u32 > /dev/null 2>&1 || echo $?",
+							if_name, get_prio(value), ht);
+		out_buf = sys_exec("tc filter add dev %s parent ffff: prio %d handle %x: protocol ip u32 divisor 256 > /dev/null 2>&1 || echo $?",
+							if_name, get_prio(value), ht);
+		if (NULL == out_buf || (out_buf[0] != '\0' && out_buf[0] != '0')) {
+			log_error("[%d] failed add ht dev %s prio %d handle %x output: %s\n",
+					pid, if_name, get_prio(value), ht, (out_buf ? out_buf : "n/a"));
+			free(cur_element);
+			rc = -EFAULT;
+			goto err;
+		}
+		out_buf = sys_exec("tc filter add dev %s protocol ip parent ffff: prio %d handle ::%x u32 ht %x:: match ip dst %s/32 hashkey mask 0x000000ff at 20 link %x: > /dev/null 2>&1 || echo $?",
+							if_name, get_prio(value), ht, ht_internal,
+							sys_ip2str(ip), ht);
+		if (NULL == out_buf || (out_buf[0] != '\0' && out_buf[0] != '0')) {
+			log_error("[%d] failed link ht dev %s prio %d handle %x:: dst %s output: %s\n",
+					pid, if_name, get_prio(value), ht, sys_ip2str(ip), (out_buf ? out_buf : "n/a"));
+			free(cur_element);
+			rc = -EFAULT;
+			goto err;
+		}
+
+		INIT_LIST_HEAD(&cur_element->list);
+		cur_element->ref = 0;
+		cur_element->value[0] = value->type;
+		cur_element->value[1] = ip;
+		cur_element->ht_id = ((ht_internal << 16) & 0xFFFF0000) | (ht & 0x0000FFFF);
+		list_add_tail(&cur_element->item, cur_head);
+	}
+	assert(cur_element);
+	cur_element->ref++;
```

Comment:
Why not increasing ref count in the end of the function? errors might happen...

Reply:
but this element was created and added into the list